File: 0_3.leetcode/java/Letter_Combinations_Of_A_Phone_Number.java
import java.util.ArrayList;
import java.util.List;

public class Letter_Combinations_Of_A_Phone_Number {
    private static final char OFFSET = '2';
    private static String[] letters = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    public List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<>();
        combinations(result, "", digits, 0);
        return result;
    }

    private void combinations(List<String> result, String current, String digits, int idx) {
        if (idx == digits.length()) {
            if (!current.isEmpty()) {
                result.add(current);
            }
            return;
        }

        String letter = letters[digits.charAt(idx) - OFFSET];
        for (int i=0;i<letter.length();i++) {
            combinations(result, current + letter.charAt(i), digits, idx + 1);
        }
    }
}

File: 0_3.leetcode/kotlin/src/main/kotlin/Letter_Combinations_Of_A_Phone_Number.kt
fun letterCombinations(digits: String): List<String> {
    val result = mutableListOf<String>()
    val offset: Char = '2'
    val letters = arrayOf("abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz")

    fun combinations(result: MutableList<String>, digits: String, current: String, idx: Int) {
        if (idx == digits.length) {
            if (current.isNotEmpty()) {
                result.add(current)
            }
            return
        }

        val letter = letters[digits[idx] - offset]
        for (ch in letter) {
            combinations(result, digits, current + ch, idx + 1)
        }
    }
    combinations(result, digits, "", 0)
    return result
}